teps/0145-cel-in-whenexpression.md

```diff
@@ -0,0 +1,372 @@
+    values: ["README.md"]
+```
+
+Current `WhenExpression` requires users to define three fields: `input`, `operator` and `values`. but CEL expressions can achieve the same functionality with a single string. Additionally, the current operator only supports `in` and `notin`, which can be limiting for complex use cases. For example, users will need to figure out [hacky workarounds][#3591] to support any in their case.
+
+For these reasons, `WhenExpressions` would be a good place to start introducing CEL in Tekton Pipelines. This would allow users to write more concise and powerful conditional expressions, and it could be extended to other fields in the future.
+
+### Use Cases
+
+- Cover all the functionalities of current `WhenExpression` and offer a simpler way of having the conditional execution
+- Support ANY: e.g. Execute the task with if any params is not empty
+- Support numeric operators such as GreaterThan/LessThan, e.g. Execute the task if test coverage is large than a certain number.
+
+### Goals
+- Support CEL in `WhenExpression`.
+
+### Non-Goals
+- Support CEL in Pipeline API fields other than `WhenExpression`.
+
+### Requirements
+- The syntax should be extensible to support other expression languages in the future if needed.
+- Users cannot inject CEL via variables such as params and results.
+
+### Related Work
+
+Similar expression evaluations are supported in other CI/CD tools in the industry:
+- [Argo][Argo]: `Workflows` support conditional execution using a when property to specify whether to run or skip a `Step` ([example][argo example]).
+- [GitHub Actions][GitHub Actions]: supports evaluating expressions in `if` of a `step` to enable conditional execution.
+- [Jenkins][Jenkins]: executes the stage when the specified Groovy expression evaluates to true.
+- [Spinnaker][Spinnaker]: Uses string expressions, `Stages` only run when expressions evaluate to `True`.
+
+
+## Proposal
+
+This doc proposes to add a `string` type field `cel` under `WhenExpression`. Like [CelCustomRun][CELCustomRun], we propose to use [cel-go][cel-go] for the expression evaluation. This will allow users to craft any valid CEL expression as defined by the [cel-spec][cel-spec] language definition.
+
+
+### Syntax
+
+In this section we will show how CEL can support the [use cases](#use-cases).
+
+#### Use CEL to replace current WhenExpression
+```yaml
+# current WhenWxpressions
+when:
+  - input: "foo"
+    operator: "in"
+    values: ["foo", “bar”]
+  - input: "duh"
+    operator: "notin"
+    values: ["foo", “bar”]
+
+# with cel
+when:
+  - cel: "'foo' in ['foo', 'bar']”
+  - cel: "'duh' not in ['foo', 'bar']”
+```
+
+**Note:** The `WhenExpressions` contains a list of `WhenExpression`, if all `WhenExpressions` are true it will return true and allow execution. So we’re using AND to evaluate the list of `WhenExpressions`. And this will remain the same for this proposal. Each existing `WhenExpression` can be replaced with a CEL expression.
+
```

Comment:
The two biggest disadvantages of `whenExpression` are (1) limited by a single operator (`in` or `notin`) for any given expression and (2) the list of expressions are evaluated using `and` operator.

With `CEL`, many conditional logic might not require more than one `whenExpressions` and might not be limited to both these shortcomings of existing `whenExpressions`.

Along with this note, its worth explicitly defining the power of `CEL` in which users might not even be worried about this.

Reply:
Thanks!!

Reply:
Added